src/serializers/encoder/v7/encode_graph.c

```diff
@@ -85,37 +152,50 @@ void RdbSaveGraph_v7(RedisModuleIO *rdb, void *value) {
 	// Acquire a read lock if we're not in a thread-safe context.
 	if(_shouldAcquireLocks()) Graph_AcquireReadLock(gc->g);
 
-	// If it is the start of the encodeing, select the phase to start from.
-	if(GraphEncodeContext_GetEncodeState(gc->encoding_context) == INIT) _SelectFirstState(gc);
-
 	// Save header
 	_RdbSaveHeader(rdb, gc);
 
-	switch(GraphEncodeContext_GetEncodeState(gc->encoding_context)) {
-	case NODES:
-		RdbSaveNodes_v7(rdb, gc);
-		break;
-	case DELETED_NODES:
-		RdbSaveDeletedNodes_v7(rdb, gc);
-		break;
-	case EDGES:
-		RdbSaveEdges_v7(rdb, gc);
-		break;
-	case DELETED_EDGES:
-		RdbSaveDeletedEdges_v7(rdb, gc);
-		break;
-	case GRAPH_SCHEMA:
-		RdbSaveGraphSchema_v7(rdb, gc);
-		break;
-	default:
-		assert(false && "Unkown encoding phase");
-		break;
+	// Save payloads info for this key and retrive the key schema.
+	PayloadInfo *key_schema = _RdbSaveKeySchema(rdb, gc);
+
+	uint payloads_count = array_len(key_schema);
+	for(uint i = 0; i < payloads_count; i++) {
+		// If the current key encoding more than one payload type, payloads count >1 and we are in a new state, zero the entities count.
+		if(i > 0) GraphEncodeContext_SetProcessedEntitiesCount(gc->encoding_context, 0);
+		PayloadInfo payload = key_schema[i];
+		switch(payload.state) {
+		case NODES:
+			RdbSaveNodes_v7(rdb, gc, payload.entities_count);
+			break;
+		case DELETED_NODES:
+			RdbSaveDeletedNodes_v7(rdb, gc, payload.entities_count);
+			break;
+		case EDGES:
+			RdbSaveEdges_v7(rdb, gc, payload.entities_count);
+			break;
+		case DELETED_EDGES:
+			RdbSaveDeletedEdges_v7(rdb, gc, payload.entities_count);
+			break;
+		case GRAPH_SCHEMA:
+			RdbSaveGraphSchema_v7(rdb, gc);
+			break;
+		default:
+			assert(false && "Unkown encoding phase");
+			break;
+		}
+		// Save the current state and the number of encoded entities.
+		GraphEncodeContext_SetEncodeState(gc->encoding_context, payload.state);
+		uint64_t offset = GraphEncodeContext_GetProcessedEntitiesCount(gc->encoding_context);
+		GraphEncodeContext_SetProcessedEntitiesCount(gc->encoding_context, payload.entities_count + offset);
 	}
+	array_free(key_schema);
 
 	// Increase processed key count. If finished encoding, resert context.
 	GraphEncodeContext_IncreaseProcessedCount(gc->encoding_context);
```

Comment:
Please rename `IncreaseProcessedCount` as it isn't meaningfully enough, what is processed?